src/x86_64/ftrace.s

```diff
@@ -0,0 +1,98 @@
+;; BK: the logic here is similar to Linux, but much simpler as we don't have
+;; to deal with fentry, dynamic tracing, and other complexities
+
+%include "frame.inc"
+
+;; save callee rip and rbp 
+%define SAVE_FRAME_SIZE (8 + 16)
+
+;; size of stack used to save mcount regs in save_mcount_regs
+%define SAVE_REG_SIZE (FRAME_MAX*8 + SAVE_FRAME_SIZE)
+
+;; size of mcount call
+%define MCOUNT_INSN_SIZE 5
+
+extern __current_ftrace_trace_fn
+extern __current_ftrace_graph_return
+
+;; After this is called, the following registers contain:
+;;  %rdi - holds rip of tracee (address of function being traced)
+;;  %rsi - holds return address of tracee
+;;  %rdx - holds the original %rbp
+;;
+;; and the stack will look like:
+;;  [            |             ]  <--- rsp
+;;  [            |             ]
+;;  [    <saved frame ctx>     ]
+;;  [            |             ]
+;;  [            |             ]
+;;  [           rbp            ]  <--- rbp
+;;  [   tracee function addr   ]  <--- moved into rdi and stored in stack(RIP)
+;;  [           rbp            ]  <--- moved into rdx and stored in stack(RBP)
+;;  [ tracee retaddr in parent ]  <--- moved into rsi 
+%macro save_mcount_regs 0
+    push rbp
+    push qword [rsp+8]
+    push rbp
+    mov rbp, rsp
+ 
+    ;; we add enough stack to save all the regs, but only need those
+    ;; needed and potentially clobbered by mcount
+    sub rsp, (SAVE_REG_SIZE - SAVE_FRAME_SIZE)
```

Comment:
I haven't worked through all this, but can we be assured that rsp always stays 16-byte aligned?

Reply:
good call, thanks